png-read-write.md

```diff
@@ -0,0 +1,63 @@
+# PNG Read/Write
+
+I only implemented parts of the reader for now so some things from bellow will definitely change a bit in order to support the writer as well.
+
+## Structure
+
+
+- wg/format/png
+	- common.d
+	- reader.d
+	- writer.d
+
+##### Common.d
+
+It contains code common for reader and writer: structs that mirror the file records or are part of both APIs, definitions of chunk types and similar. The main API data structure is this:
+
+```d
+struct Result(T) {
+    string errorMessage;
+    T data;
+}
+
+Result!PngHeaderData loadPngHeader(const ubyte[] data);
+Result!ImageBuffer loadPng(const ubyte[] data, Allocator* allocator)
+Result!ImageBuffer loadPng(const ubyte[] data, Allocator* allocator, Allocator* chunkAllocator, out PngChunks* chunks)
+
+
+// This structure mirrors the file record
+align(1)
+struct PngHeaderData
+{
+    uint width;
+    uint height;
+    PngBitDepth bitDepth;
+    PngColorType colorType;
+    PngCompressionMethod compressionMethod;
+    PngFilterMethod filterMethod;
+    PngInterlaceMethod interlaceMethod;
+}
+```
+
+##### Reader.d
+
+With the above API you can read just the headers or the whole image. The ImageBuffer would contain raw pixel data from the file and all chunks including the IHDR would be included in metadata. In case raw pixel data is in color type 3 (palette indices), it would actually be depalettized into RGB or RGBA if transparency chunk is also present but those two chunks would still be present in the metadata. The ImageBuffer format could thus end up being:
+
+- l, l_1, l_2, l_4, l_16 for grayscale images without alpha and without transparency
+- la, la_1_1, la_2_2, la_4_4, la_16_16 for grayscale images with alpha or with transparency
+- rgb, rgb_16 for palette images without transparency and rgb images
+- rgba, rgba_16 for palette images with transparency and rgba images
+
+Any of these might get "^<g>" where <g> is non standard gamma read from the gamma chunk and "R{x,y,Y}G{x,y,Y}B{x,y,Y}" if there is a cHRM chunk. I have no idea what, if anything, should the reader do with sRGB and iCCP chunks if they are present.
```

Comment:
sRGB is the default assumption for colourspace, so if that chunk is present, then leave the colourspace as default (ie, not specified). iCCP needs a decoder, it will likely want to populate a metadata chunk with very comprehensive information... just ignore it for now (assume sRGB?)

Reply:
ICC profile support is a very advanced use case

Reply:
Indeed, ICC is a mini-library on its own.